### ISSUE TITLE
Running Mapper on Batches 

### ISSUE THREAD

**Issue by omermosa** [oneThousand1000/HairMapper#15]
Dear all,
Is there a way to paralyze the mapper code? i.e., if I have a batch of latent codes, is it possible to run the line below on a batch?
`mapper(mapper_input_tensor)`
Thanks!

**Comment by oneThousand1000** [comment]
Hi omermosa!

You can input a batch of latent codes to HairMapper. But there are some modifications:

In [single_mapper.py](https://github.com/oneThousand1000/HairMapper/blob/master/mapper/networks/single_mapper.py), Line 40

```
         x = bias_act.bias_act(x, b, act=self.activation,dim=2) # x = bias_act.bias_act(x, b, act=self.activation)
```
Line 82-91

```
        #x = x.squeeze(0)
        # Main layers.
        for idx in range(self.num_layers):
            layer = getattr(self, f'fc{idx}')
            x = layer(x)  # W
        #W_latent = torch.clone(x.unsqueeze(0))

        return x
```


then you can input a batch of latent codes to HairMapper like this:


```
    codes = []
    for code_path in code_list:
        code = np.load(code_path)
        codes.append(code)
    codes = np.concatenate(codes, axis=0) # shape: [batch size, 18, 512]

    mapper_input = codes.copy()
    mapper_input_tensor = torch.from_numpy(mapper_input).cuda().float()
    print('mapper_input_tensor shape: ', mapper_input_tensor.shape)
    edited_latent_codes_debug = codes.copy()
    mapper_out = mapper(mapper_input_tensor)
    print('mapper_out shape: ', mapper_out.shape, 'alpha', alpha)

    edited_latent_codes_debug[:, :8, :] += alpha * mapper_out.to('cpu').detach().numpy()

```
Good luck!

**Comment by omermosa** [comment]
That is great; thank you so much!
I guess there is nothing to do to make the rest of the operations (getting hair mask and seamless clone ) run on batches right? since they involve opencv operations,

**Comment by oneThousand1000** [comment]
Yes, it is not easy to run the remaining operations on batches. This is also the reason why, during testing, we only process one image at a time.

**Comment by omermosa** [comment]
Yes, makes sense. Thank you very much!
